src/ir_visitor/visitor.rs

```diff
@@ -287,19 +226,24 @@ where
         }
 
         if let Some(pre_cmds) = pre_cmds {
-            data.comp.cmds = pre_cmds;
+            data.comp().cmds = pre_cmds;
         }
-        data.comp.cmds.extend(cmds);
+        data.comp().cmds.extend(cmds);
 
         self.end(&mut data);
     }
 
     /// Apply the pass to all components in the context
-    fn do_pass(opts: &cmdline::Opts, ctx: &mut ir::Context) -> Result<(), u64> {
+    fn do_pass<'pass: 'comp>(
+        opts: &'pass cmdline::Opts,
+        ctx: &'pass mut ir::Context,
+    ) -> Result<(), u64> {
         let mut visitor = Self::from(opts, ctx);
         for idx in ctx.comps.idx_iter() {
             visitor.clear_data();
-            visitor.visit((idx, opts, &mut *ctx).into());
+            let opts: &'comp cmdline::Opts = opts;
+            let ctx: &'comp mut ir::Context = &mut *ctx;
+            visitor.visit(Self::Data::new(idx, opts, ctx));
```

Comment:
Is there any way to do what I'm trying to here? Been spending way too long on it and making no progress...

Reply:
The [lifetime bounds](https://doc.rust-lang.org/beta/reference/trait-bounds.html#lifetime-bounds) syntax is a bit confusing but your code currently says that `'pass` outlives `'comp` which is probably the opposite of what you want.

Reply:
I actually want `'pass` to outlive `'comp` - `comp` should be alive for each component (each iteration of the loop) while `'pass` should be a reference to the context that stays alive for the entire pass.